### DISCUSSION TITLE
Wagmi React hooks - caching

### DISCUSSION THREAD

**Discussion by Cezarry** [wevm/wagmi#1988] [Q&A]
I wonder if someone use Wagmi hooks for Ethereum and it's caching mechanism. Since docs doesn't say much about it ([Wagmi docs](https://wagmi.sh/react/hooks/useContractRead#cachetime-optional)) can somebody explain me how it acutualy works? If i use `useContractRead` with `scopeKey: 'myScope'` and caching `cacheTime: 1000*60*10 // 10 minutes` then every time my component mounts it should first read the cache and return a cached value instead of using RPC to get value from blockchain, right? Well if it should, then it doesn't work, because every time I switch page I can see that my local hardhat is being asked for values. This means it can drain out RPC query limit very fast. How should I use cacheTime? Maybe I should also use `staleTime`?

**Comment by Cezarry** [comment]
Anyone?

**Reply by 0xbuidler** [reply]
Have you find an answer yet ?

**Reply by Cezarry** [reply]
> Have you find an answer yet ?

Unfortunately no :/

**Reply by 0xbuidler** [reply]
People seems to say that caching works with hooks, like `useReadContract` but i'm wondering if it's the same for simple functions from WAGMI core like`readContract` which does the same but not with a hook.

**Reply by Cezarry** [reply]
> People seems to say that caching works with hooks, like `useReadContract` but i'm wondering if it's the same for simple functions from WAGMI core like`readContract` which does the same but not with a hook.

Then it doesn't work for me. Even if I set long caching time when I reload page it consumes data from blockchain instead of cache (I can see it in hardhat node that every time query goes to it).

**Reply by 0xbuidler** [reply]
Yeah I know exactly what you are talking about because it happens to me as well. I'm a bit concerned tbh. I'll try to ping one of the main contributors. @tmm @jxom . Hopefully they can help us. I'm planning on deploying to testnet/mainnet soon and I need insights.

**Reply by jxom** [reply]
wagmi will read from the cache first, and do a background fetch. If you want to read from cache only, you will need to utilise `staleTime`.

The `cacheTime` option is the time in which the data remains in the cache.
The `staleTime` option is the time in which the cached data should be served without performing a background fetch.

wagmi uses React Query for this functionality internally - you can read more on these options here: https://tkdodo.eu/blog/practical-react-query#the-defaults-explained

**Reply by b-pmcg** [reply]
Hey @jxom in the article you linked, the default `cacheTime` for react-query is 5 minutes, but for wagmi the default is 0. Just wondering why wagmi chose that, and will the defaults be the same for the upcoming v1 & v2 releases?
